src/shared/varlink.c

```diff
@@ -543,9 +589,31 @@ static int varlink_read(Varlink *v) {
 
         bool prefer_read = v->prefer_read_write;
         if (!prefer_read) {
-                n = recv(v->fd, v->input_buffer + v->input_buffer_index + v->input_buffer_size, rs, MSG_DONTWAIT);
+                struct msghdr msgh;
+                struct iovec iov;
+                CMSG_BUFFER_TYPE(CMSG_SPACE(sizeof(struct ucred))) cmsg = {};
+                struct ucred *ucred;
+
+                iov = IOVEC_MAKE(v->input_buffer + v->input_buffer_index + v->input_buffer_size, rs);
+                msgh = (struct msghdr) {
+                        .msg_iov = &iov,
+                        .msg_iovlen = 1,
+                        .msg_control = cmsg.buf,
+                        .msg_controllen = sizeof(cmsg.buf),
+                };
+
+                n = recvmsg_safe(v->fd, &msgh, MSG_DONTWAIT);
                 if (n < 0 && errno == ENOTSOCK)
                         prefer_read = v->prefer_read_write = true;
+
+                if (n != -1) {
+                        ucred = CMSG_FIND_DATA(&msgh, SOL_SOCKET, SCM_CREDENTIALS, struct ucred);
+                        if (ucred) {
+                                v->ucred_recv = *ucred;
+                                v->ucred_recv_acquired = true;
+                        }
+                        log_debug("Passed creds %s acquired", v->ucred_recv_acquired ? "successfully" : "could not be");
```

Comment:
please remove or downgrade to log_trace(). this will otherwise be written on each read, which might be a bit too often. 

also, formatting literal strings via format strings is weird. maybe just make this two log_trace() calls, and move one into the if block above, and one to an else branch